Add Authorization in Kubernetes Hardening Guide

ashish493 opened the pull request:
This is Authorization Section of Kubernetes hardening guide which has been under discussion in SIG-Security-Docs https://github.com/kubernetes/sig-security/issues/30 .

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+authorization policy is an important step in preventing unauthorized access to the clusters or privilege escalation attacks. Kubernetes offers
+multiple ways to enforce permissions which are represented by various authorization modes and modules - Node authorization, ABAC, Webhook,
+RBAC. Using the RBAC mode is the most standard way of designing permissions for cluster users and the workloads. We can set the authorization
+mode using the `--authorization-mode` flag in kube api server. A proper Authorization design helps in reducing the blast radius of any kind of
```
```suggestion
mode using the `--authorization-mode` flag in the API server. A proper Authorization design helps in reducing the blast radius of any kind of
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+RBAC mode is enabled by default and is one method to control access to cluster resources based on the roles of individuals within an
+organization. It is the most important authorization method for both developers and admins in Kubernetes. It is used to restrict access for
+user accounts and service accounts. To check if RBAC is enabled in a cluster using kubectl, execute `kubectl api-version`. The RBAC mode can be
+enabled by setting the RBAC value with --authorizationmode flag in kube api server command.
```
```suggestion
enabled by setting `--authorization-mode` flag to "RBAC" in the `kube-apiserver` command line.
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+user accounts and service accounts. To check if RBAC is enabled in a cluster using kubectl, execute `kubectl api-version`. The RBAC mode can be
+enabled by setting the RBAC value with --authorizationmode flag in kube api server command.
+
+> `kube-apiserver --authorization-mode=RBAC`
```
````suggestion
```
kube-apiserver --authorization-mode=RBAC
```
````

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+> `kube-apiserver --authorization-mode=RBAC`
+
+### Least Privilege Principal
```
```suggestion
### Least Privilege Principle
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+### Least Privilege Principal
+
+- The principal of least privilege should always be followed while designing the authorization policies. This principle states that any user or
```
```suggestion
- The principle of least privilege should always be followed while designing the authorization policies. This principle states that any user or
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+- Each group should only have access to the resources that they need and should not be allowed to view or alter the resources of other groups.
+  Users, user groups, and service accounts should only be able to interact with and access resources within certain namespaces.
+
+- We should never use _AlwaysAllow_ value in the authorization-mode flags, since it basically disables all the authorization modes, limiting
```
```suggestion
- We should never use "AlwaysAllow" value in the `--authorization-mode` flag, since it basically disables all the authorization modes, limiting
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+- We should never use _AlwaysAllow_ value in the authorization-mode flags, since it basically disables all the authorization modes, limiting
+  the ability to enforce the least privilege for access.
+
+- Using the _system:masters_ group for assigning rights should always be avoided, since this group has hard-coded cluster-admin rights which
```
```suggestion
- Using the `system:masters` group when assigning rights should always be avoided, since this group has hard-coded cluster-admin rights which
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+- Using the _system:masters_ group for assigning rights should always be avoided, since this group has hard-coded cluster-admin rights which
+  cannot be revoked (it’s hardcoded in the source code). Any user having this right will always be admin. A recommendation would be that if
+  people really need cluster-admin, then just provide a binding to the cluster-admin clusterrole.
```
```suggestion
  people really need cluster-admin, then just provide a binding to the `cluster-admin` clusterrole.
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+### Resources to Restrict to prevent Privilege Escalation
+
+- **_Etcd_** is a critical component which stores state information and cluster-secrets in key-value pairs in the database. Unauthorized access
```
```suggestion
- **_Etcd_** is a critical component which stores state information and cluster Secrets in key-value pairs in the database. Unauthorized access
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+  Kubernetes admission control.
+
+- **_Kubernetes Dashboard_** is a web-based UI to interact with applications running in the cluster. Attackers with elevated access to the
+  dashboard can open a shell connection to any pod, view secrets and can also use the existing resources for cryptojacking attacks. In the
```
```suggestion
  dashboard can open a shell connection to any pod, view Secrets and can also use the existing resources for cryptojacking attacks. In the
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+  cluster role, users should only be provided with the read-only permission to the dashboard and write access should be minimized at all costs.
+
+- **_Kubernetes Secret_** is an object that contains sensitive information, such as a password, token, or key, which is used in the Pod
+  configuration or container image. Unrestricted access of secrets to the components, which doesn't require them expands the attack surface. We
```
```suggestion
  configuration or container image. Unrestricted access of Secrets to the principals which don't require them expands the attack surface. We
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+- **_Kubernetes Secret_** is an object that contains sensitive information, such as a password, token, or key, which is used in the Pod
+  configuration or container image. Unrestricted access of secrets to the components, which doesn't require them expands the attack surface. We
+  can use admission controllers to define policies to restrict the use of secrets only to the required components.
```
```suggestion
  can use admission controllers to define policies to restrict the access of Secrets only to the required components.
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+  can use admission controllers to define policies to restrict the use of secrets only to the required components.
+
+- **_Kubernetes API_** is an HTTP API used to query and modify the status of Kubernetes objects. It also facilitates communication between
+  users, various components inside cluster and external parties. Unrestricted access to this API may lead to resource modifications, data
```
```suggestion
  users, various components inside the cluster and the external parties. Unrestricted accesses to this API may lead to resource modifications, data
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+
+### Admission Controllers
+
+We can extend the in built RBAC policies using the Validation Webhook of Admission Controllers to strengthen the authorization design. There
```
```suggestion
We can extend the builtin RBAC policies using the validation admission webhooks to strengthen the authorization design. There
```

tengqm commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,88 @@
+approve them or deny them. The request gets evaluated after it has been verified and authorized by the API server before it is granted and
+implemented. This is an optional feature, which may be only be required for large scale clusters or where complex security is required. They
+can be adjustable for many different user-specific scenarios and environments. There are many open source and commercial implementations from
+which organizations can choose and enforce their specific restrictions.
```
Move this paragraph ahead of the previous one?

ashish493 replied:
Yup, moved it above.

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+  cannot be revoked (it’s hardcoded in the source code). Any user having this right will always be admin. A recommendation would be that if
+  people really need cluster-admin, then just provide a binding to the `cluster-admin` clusterrole.
+
+### Resources to Restrict to prevent Privilege Escalation
```
```suggestion
### Restriction of resources for prevention of Privilege Escalation
```

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+  cluster role, users should only be provided with the read-only permission to the dashboard and write access should be minimized at all costs.
+
+- **_Kubernetes Secret_** is an object that contains sensitive information, such as a password, token, or key, which is used in the Pod
+  configuration or container image. Unrestricted access of Secrets to the principals which don't require them expands the attack surface. We
```
```suggestion
  configuration or container image. Unrestricted access of Secrets to the principals which don't require them, expands the attack surface. We
```

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+### Admission Controllers
+
+We can extend the builtin RBAC policies using the validation admission webhooks to strengthen the authorization design. There
```
```suggestion
We can extend the built-in RBAC policies using the validation admission webhooks to strengthen the authorization design. There
```

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+A Kubernetes admission controller is a component of code that analyzes requests made to the Kubernetes API server and decides whether to
+approve them or deny them. The request gets evaluated after it has been verified and authorized by the API server before it is granted and
+implemented. This is an optional feature, which may be only be required for large scale clusters or where complex security is required. They
```
```suggestion
implemented. This is an optional feature, which maybe only required for large scale clusters or where complex security is required. They
```

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+A Kubernetes admission controller is a component of code that analyzes requests made to the Kubernetes API server and decides whether to
+approve them or deny them. The request gets evaluated after it has been verified and authorized by the API server before it is granted and
+implemented. This is an optional feature, which may be only be required for large scale clusters or where complex security is required. They
+can be adjustable for many different user-specific scenarios and environments. There are many open source and commercial implementations from
```
```suggestion
can be adjusted for different user-specific scenarios and environments. There are many open source and commercial implementations from
```

aj11anuj commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+As the application grows, the scope of the components expands and the required permissions of certain resources also changes. There is a very
+high chance that any admin or developer might unknowingly give extra permission to any of the defined roles any time. Regular reviews of the
+existing RBAC policies are required to check and prevent any such. We can use the below tools to audit and make a check for risky permissions
```
```suggestion
existing RBAC policies are required to check and prevent such mistakes. We can use the below tools to audit and make a check for risky permissions
```

aj11anuj reviewed (changes requested):
Some minor suggestions from my side

ashish493 commented:
@aj11anuj Thanks, I've made some changes by checking with grammarly.

divya-mohan0209 commented:
/remove-lifecycle stale

ashish493 commented:
@divya-mohan0209 Yeah I'm currently working on the Resource Limits topic for the Hardening Guide. Thanks for the pointing out the [Kubernetes style guide](https://kubernetes.io/docs/contribute/style/style-guide/) . I will make the changes in this doc according to it.

divya-mohan0209 commented:
Hello @ashish493, checking in on the updates for the PR. Is this still being worked upon?

ashish493 commented:
Hi @divya-mohan0209 
Yeah, I've made the changes locally and will update this PR within this weekend. Sorry for the delay, was occupied in the release for my company.

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+<!-- overview -->
+
+Authorization in Kubernetes ecosystem means specifying the privileges/permissions for a particular kubernetes resource. Defining a proper
```
nit: grammar and 
```suggestion
Authorization in the Kubernetes ecosystem means specifying the privileges/permissions for a particular Kubernetes resource. Defining a proper
```

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+<!-- overview -->
+
+Authorization in Kubernetes ecosystem means specifying the privileges/permissions for a particular kubernetes resource. Defining a proper
+authorization policy is an important step in preventing unauthorized access to the clusters or privilege escalation attacks. Kubernetes offers
```
If this sentence is referring to a single cluster then I suggest the following
```suggestion
authorization policy is an important step in preventing unauthorized access to the cluster or privilege escalation attacks. Kubernetes offers
```

If this sentence is referring to multiple clusters then I suggest
```suggestion 
authorization policy is an important step in preventing unauthorized access to clusters or privilege escalation attacks. Kubernetes offers
```

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+Authorization in Kubernetes ecosystem means specifying the privileges/permissions for a particular kubernetes resource. Defining a proper
+authorization policy is an important step in preventing unauthorized access to the clusters or privilege escalation attacks. Kubernetes offers
+multiple ways to enforce permissions which are represented by various authorization modes and modules - Node authorization, ABAC, Webhook,
```
non-blocking nit
```suggestion
multiple ways to enforce permissions which are represented by various authorization modes and modules - Node authorization, ABAC, webhook,
```

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+Authorization in Kubernetes ecosystem means specifying the privileges/permissions for a particular kubernetes resource. Defining a proper
+authorization policy is an important step in preventing unauthorized access to the clusters or privilege escalation attacks. Kubernetes offers
+multiple ways to enforce permissions which are represented by various authorization modes and modules - Node authorization, ABAC, Webhook,
+RBAC. Using the RBAC mode is the most standard way of designing permissions for cluster users and the workloads. We can set the authorization
```
would links be helpful?
```suggestion
multiple ways to enforce permissions which are represented by various authorization modes and modules -[ Node authorization](/docs/reference/access-authn-authz/node/), [ABAC](/docs/reference/access-authn-authz/abac/), [Webhook](/docs/reference/access-authn-authz/webhook/),
[RBAC](/docs/reference/access-authn-authz/rbac/). Using the RBAC mode is the most standard way of designing permissions for cluster users and the workloads. We can set the authorization
```

ashish493 replied:
Yup sounds good.

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+- **_Etcd_** is a critical component which stores state information and cluster Secrets in key-value pairs in the database. Unauthorized access
+  to Etcd can compromise the entire cluster. Roles for the users with only the required access to certain keys should be set.
+
+- **_Kubelet_** is a primary node agent which operates on every node. The kubelet service should be run with `--anonymous-auth=false`. The
```
kubelet is typically spelled lower case
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/

```suggestion
- **_kubelet_** is a primary node agent which operates on every node. The kubelet service should be run with `--anonymous-auth=false`. The
```

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+### Resources to Restrict to prevent Privilege Escalation
+
+- **_Etcd_** is a critical component which stores state information and cluster Secrets in key-value pairs in the database. Unauthorized access
```
etcd is typically spelled lower case
https://kubernetes.io/docs/concepts/overview/components/#etcd

```suggestion
- **_etcd_** is a critical component which stores state information and cluster Secrets in key-value pairs in the database. Unauthorized access
```

reylejano commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,89 @@
+
+As the application grows, the scope of the components expands, and the required permissions for certain resources also change. There is a very
+There is a high chance that any admin or developer might unknowingly give extra permission to any of the defined roles at any time. Regular reviews of the
+existing RBAC policies are required to check for and prevent such mistakes. We can use the below tools to audit and make a check for risky permissions
+in the existing RBAC policies defined in the cluster.
+
+- [K8s-rbac-audit](https://github.com/cyberark/kubernetes-rbac-audit)
+- [Krane](https://github.com/appvia/krane)
+- [Kubiscan](https://github.com/cyberark/KubiScan)
```
If we suggest tools, it may seem that the project is picking "the best" audit tools.
I think we can leave the links to the tools out.
```suggestion
existing RBAC policies are required to check for and prevent such mistakes. 
```
According to the [style guide](https://kubernetes.io/docs/contribute/style/content-guide/#third-party-content):

> Docs can link to third-party open source software (OSS) outside the Kubernetes project only if it's necessary for Kubernetes to function

or 

```suggestion
existing RBAC policies are required to check for and prevent such mistakes. There are open source tools available to audit the cluster's RBAC policies.
```

ashish493 commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
````diff
@@ -0,0 +1,89 @@
+kube-apiserver --authorization-mode=RBAC
+```
+
+### Least Privilege Principle
````
Sure, will check and update!

ashish493 commented:
Thanks @reylejano for reviewing, I've updated the changes according to it. 

@divya-mohan0209 I've updated the content according to [Kubernetes style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,60 @@
+
+<!-- body -->
+
+### Role based access control(RBAC)
```
```suggestion
## Role based access control (RBAC) {#hardening-rbac}
```

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,60 @@
+- It is crucial to refrain from using the AlwaysAllow value in the --authorization-mode flag. This value effectively disables all authorization modes, compromising the ability to enforce the principle of least privilege.
+- Assigning rights using the `system:masters` group should always be avoided, as it possesses hardcoded cluster-admin rights. These rights cannot be revoked due to their hardcoded nature in the source code. Users with `system:masters` rights will always have administrative privileges. The groups who genuinely require cluster-admin rights should be provided a binding to the cluster-admin clusterro ...
+
+### Resources to restrict to prevent privilege escalation
```
```suggestion
## Resources to restrict to prevent privilege escalation {#hardening-components}
```

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,60 @@
+
+To prevent privilege escalation, specific measures should be taken to restrict access to critical resources within the Kubernetes environment:
+
+- *etcd*: A crucial component storing state information and cluster Secrets, requires careful access control. Roles should be defined for users with access limited to specific keys to prevent unauthorized access, safeguarding the entire cluster.
```
Make it clear that these aren't roles in the RBAC sense

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```diff
@@ -0,0 +1,60 @@
+
+- *etcd*: A crucial component storing state information and cluster Secrets, requires careful access control. Roles should be defined for users with access limited to specific keys to prevent unauthorized access, safeguarding the entire cluster.
+
+- *kubelet*: The primary node agent operating on every node, necessitates secure configuration. The kubelet service should run with --anonymous-auth=false to enhance security. The node/proxy right should be granted judiciously, ensuring only necessary users have access to the Kubelet API and preventing evasion of Kubernetes admission control.
```
- Put command line arguments inside backticks.

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
Ideally, reword this to be good advice whether or not the cluster in question uses Kubernetes RBAC.

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
Ideally, reword this to be good advice whether or not the cluster in question uses Kubernetes RBAC.

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```suggestion
## Admission control {#hardening-admission-control}
```

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
This is poor advice; if you want custom authz, you should use an **authorization** callout: https://kubernetes.io/docs/reference/access-authn-authz/webhook/

sftim commented on content/en/docs/concepts/security/hardening-guide/authorization-hardening.md:
```suggestion
## Policy review
```

sftim reviewed (changes requested):
Thanks @ashish493 

We've got some quite good documentation already, which sets a bar for quality. I'd only be happy to merge this if you:
- remove the implication that RBAC mode is always active
- fix the snag around recommending admission controllers, rather than authz webhooks
  (it's OK to mention admission control as further reading)

and please also consider
- fix the wrapping for the Markdown source, so that lines are typically 100 characters of fewer

divya-mohan0209 commented:
@ashish493 : Please could you advise if you're able to address the suggestions made by @sftim in [this review](https://github.com/kubernetes/website/pull/43623#pullrequestreview-1926618339)?

ashish493 commented:
@divya-mohan0209 need some more time for those changes. We have sig-security-docs meeting this thursday. I will ask for some help for proceeding further. After that I will update my progress here.

dipesh-rawat commented:
@ashish493 What's the plan for this PR moving forward? Does it require more refinement, is it ready for review _(pending [review comments](https://github.com/kubernetes/website/pull/43623#pullrequestreview-1926618339))_, or are we at the stage where this can be closed?

ashish493 commented:
@dipesh-rawat Yeah, it needs a little refinement. I couldn't give much time to it because of some personal issues. I will definitely try to do it soon. If it comes to a point where I won't be able to do it, then I will definitely pass this on to sig-security-docs for further enhancement.

divya-mohan0209 commented:
@ashish493 Thank you for your efforts! Since there has been no progress on this PR, I'll be closing it out. In the event that you'd like to pass it on to SIG Security Docs and reopen it, please feel free to do so whenever you have the bandwidth.

/close
